File: Data Structures & Algorithms/search-for-word/submission-2.cpp
class Solution {
public:
    set<pair<int, int>> index;
    int ROWS, COLS;
    bool exist(vector<vector<char>>& board, string word) {
        ROWS = board.size();
        COLS = board[0].size();
        for(int i = 0; i < ROWS; i++){
            for(int j = 0; j < COLS; ++j){
                if(dfs(board, i, j, 0, word)){
                    return true;
                }
            }
        }
        return false;
    }
    bool dfs(vector<vector<char>> & board, int row, int col, int curr, string word){
        if(curr == word.length()) return true;
        if(row<0||col<0||row>=ROWS||col>=COLS||word[curr]!=board[row][col]||index.count({row, col})) return false;
        index.insert({row, col});
        bool res = dfs(board, row+1, col, curr+1, word)||
                   dfs(board, row, col-1, curr+1, word)||
                   dfs(board, row, col+1, curr+1, word)||
                   dfs(board, row-1, col, curr+1, word);
        index.erase({row, col});
        return res;
    }
};

File: Data Structures & Algorithms/surrounded-regions/submission-2.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        row, col = len(board), len(board[0])
        def dfs(i, j):
            if i >= row or j >= col or i < 0 or j < 0 or board[i][j]!='O':
                return
            board[i][j]='#'
            dfs(i+1, j)
            dfs(i-1, j)
            dfs(i, j+1)
            dfs(i, j-1)
        for i in range((row)):
            for j in range((col)):
                if (i in [0, row-1] or j in [0, col-1]) and board[i][j]=='O':
                    dfs(i, j)
        for i in range((row)):
            for j in range((col)):
                if board[i][j]=='O':
                    board[i][j]='X'
        for i in range((row)):
            for j in range((col)):
                if board[i][j]=='#':
                    board[i][j]='O'

        